File: smartedu-backend/src/main/java/com/smartedu/controller/AttendanceController.java
package com.smartedu.controller;

import com.smartedu.model.Attendance;
import com.smartedu.service.AttendanceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@CrossOrigin(origins = "*", maxAge = 3600)
@RestController
@RequestMapping("/api/attendance")
public class AttendanceController {

    @Autowired
    private AttendanceService attendanceService;

    @PostMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Attendance> markAttendance(@RequestBody Attendance attendance) {
        return ResponseEntity.ok(attendanceService.markAttendance(attendance));
    }

    @GetMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<List<Attendance>> getAllAttendance() {
        return ResponseEntity.ok(attendanceService.getAllAttendance());
    }

    @GetMapping("/student/{studentId}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER') or hasRole('STUDENT')")
    public ResponseEntity<List<Attendance>> getAttendanceByStudent(@PathVariable Long studentId) {
        org.springframework.security.core.Authentication auth = org.springframework.security.core.context.SecurityContextHolder.getContext().getAuthentication();
        boolean isStudent = auth.getAuthorities().stream().anyMatch(a -> a.getAuthority().equals("ROLE_STUDENT"));
        if (isStudent) {
            return ResponseEntity.ok(attendanceService.getApprovedAttendanceByStudent(studentId));
        }
        return ResponseEntity.ok(attendanceService.getAttendanceByStudent(studentId));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Attendance> updateAttendance(@PathVariable Long id, @RequestBody Attendance attendanceDetails) {
        return ResponseEntity.ok(attendanceService.updateAttendance(id, attendanceDetails));
    }

    @PutMapping("/{id}/approve")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Attendance> approveAttendance(@PathVariable Long id) {
        return ResponseEntity.ok(attendanceService.updateApprovalStatus(id, com.smartedu.model.RecordStatus.APPROVED));
    }

    @PutMapping("/{id}/reject")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Attendance> rejectAttendance(@PathVariable Long id) {
        return ResponseEntity.ok(attendanceService.updateApprovalStatus(id, com.smartedu.model.RecordStatus.REJECTED));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<?> deleteAttendance(@PathVariable Long id) {
        attendanceService.deleteAttendance(id);
        return ResponseEntity.ok().build();
    }
}

File: smartedu-backend/controllers/attendanceController.js
const attendanceService = require('../services/attendanceService');

async function markAttendance(req, res, next) {
    try {
        const attendance = await attendanceService.markAttendance(req.body, Boolean(req.user?.isDemo));
        return res.json(attendance);
    } catch (error) {
        next(error);
    }
}

async function getAllAttendance(req, res, next) {
    try {
        const attendanceList = await attendanceService.getAllAttendance(Boolean(req.user?.isDemo));
        return res.json(attendanceList);
    } catch (error) {
        next(error);
    }
}

async function getAttendanceByStudent(req, res, next) {
    try {
        const { studentId } = req.params;
        const userRole = req.user.role ? req.user.role.toUpperCase().replace(/^ROLE_/, '') : '';
        const isDemo = Boolean(req.user?.isDemo);
        if (userRole === 'STUDENT') {
            const attendanceList = await attendanceService.getApprovedAttendanceByStudent(studentId, isDemo);
            return res.json(attendanceList);
        }
        const attendanceList = await attendanceService.getAttendanceByStudent(studentId, isDemo);
        return res.json(attendanceList);
    } catch (error) {
        next(error);
    }
}

async function updateAttendance(req, res, next) {
    try {
        const { id } = req.params;
        const attendance = await attendanceService.updateAttendance(id, req.body, Boolean(req.user?.isDemo));
        return res.json(attendance);
    } catch (error) {
        next(error);
    }
}

async function approveAttendance(req, res, next) {
    try {
        const { id } = req.params;
        const attendance = await attendanceService.updateApprovalStatus(id, 'APPROVED', Boolean(req.user?.isDemo));
        return res.json(attendance);
    } catch (error) {
        next(error);
    }
}

async function rejectAttendance(req, res, next) {
    try {
        const { id } = req.params;
        const attendance = await attendanceService.updateApprovalStatus(id, 'REJECTED', Boolean(req.user?.isDemo));
        return res.json(attendance);
    } catch (error) {
        next(error);
    }
}

async function deleteAttendance(req, res, next) {
    try {
        const { id } = req.params;
        await attendanceService.deleteAttendance(id, Boolean(req.user?.isDemo));
        return res.json({ success: true, message: 'Attendance deleted successfully' });
    } catch (error) {
        next(error);
    }
}

module.exports = {
    markAttendance,
    getAllAttendance,
    getAttendanceByStudent,
    updateAttendance,
    approveAttendance,
    rejectAttendance,
    deleteAttendance
};
